### PR TITLE
Upgrade rich text no mdx

### LINKED ISSUE #2881: Markdown shortcodes in rich-text field cause a collection to blow up.
## Description

We're close'ish to our rich-text editor being able to support .md content, but it blows up on reading .md files with shortcodes:

```
Checking collection docs
[
  GraphQLError [Object]: Could not parse expression with acorn: Unexpected token
      at locatedError (/Users/jamesohalloran/Documents/forestry/tinacms-site-next/node_modules/graphql/error/locatedError.js:29:10)
      at /Users/jamesohalloran/Documents/forestry/tinacms-site-next/node_modules/graphql/execution/execute.js:479:52
```

I know I could convert all of my .md files to .mdx, and turn the shortcodes into MDX components, but if shortcodes were even workable as plain text, it would save a lot of time migrating existing content

### Steps to reproduce

Add "invalid" content to a`.md` file in a collection. E.g:

```md
{{ WarningCallout text="" }}
```

Make sure the body is set up with the rich-text field

```
{
        type: "rich-text",
        name: "body",
        label: "body",
        isBody: true,
}
```

Try and navigate to that collection in the CMS (or run `yarn run tinacms audit`)

### Expected result

I should be able to edit the collection's markdown content in the cms, and edit the shortcode as plain text.

### Actual result

The collection list blows up.

See [related issue](https://github.com/tinacms/tinacms/issues/2880) on how the error is actually presented

### Environment

Environment
Tinacms Package Versions
├── @tinacms/auth@0.50.2
├── @tinacms/cli@0.60.15
├── @tinacms/graphql@0.60.0
├── next-tinacms-cloudinary@3.5.12
├── react-tinacms-editor@0.53.9
├── tinacms@0.68.3

### LINKED ISSUE #2884: Links nested within "**" or "_" are giving the error: 'Not sure, this should be flattened to the same node'
## Description

It seems not possible to have links in bold or italic texts.

I.e:

`_Quis semper vulputate [aliquam](https://link.com) venenatdis egestas sagittis quisque orci._`
or
`**Quis semper vulputate [aliquam](https://link.com) venenatdis egestas sagittis quisque orci.**`

Give the error message:

```
Warning: There was an error when fetching data
Error: Unable to fetch, errors: 
        Not sure, this should be flattened to the same node
    at LocalClient.request (/Users/augurk/projects/sandbox/tina-cloud-starter/node_modules/tinacms/dist/index.js:2273:15)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async requester1 (webpack-internal:///./.tina/__generated__/types.ts:376:16)
    at async getStaticProps (webpack-internal:///./pages/posts.tsx:59:23)
    at async Object.renderToHTML (/Users/augurk/projects/sandbox/tina-cloud-starter/node_modules/next/dist/server/render.js:387:20)
    at async doRender (/Users/augurk/projects/sandbox/tina-cloud-starter/node_modules/next/dist/server/next-server.js:1428:38)
    at async /Users/augurk/projects/sandbox/tina-cloud-starter/node_modules/next/dist/server/next-server.js:1523:28
    at async /Users/augurk/projects/sandbox/tina-cloud-starter/node_modules/next/dist/server/response-cache.js:63:36
error - pages/posts.tsx (10:42) @ HomePage
```

### Steps to reproduce

You can test it with the Tina cloud starter repo.

Go to a post and try to make a couple words with a link in it bold or italic in the rich text WYSIWYG. You will get the error when you try to save it.

### Expected result

What should happen?

It's valid markdown, so it should just get parsed.

#### Tinacms Package Versions

latest

closes https://linear.app/tina/issue/ENG-115

### PR DESCRIPTION
**Pull request by jeffsee55**
Updates to the `rich-text` component as well the shape of the `rich-text` field response from the API. This will be a minor version bump.

* Adds support for `isTitle` on MDX elements
* Fixes issues related to nested marks
* Uses monaco editor for code blocks
* Improves styling of nested list items
* Improves handling of rich-text during reset
* No longer errors on unrecognized JSX/html, instead falls back to print `No component provided for <compnonent name>`
* No longer errors on markdown parsing errors, instead falls back to rendering markdown as a string, customizable via the TinaMarkdown component (`invalid_markdown` prop)
* Prepares rich-text component for raw mode - where you can edit the raw markdown directly in the Tina form. This will be available in future release. NOTE: there are some areas where the code is just commented out.

---

Apologies for the messy commits, I should have rebased before merging main! There are 2 areas of focus for this code:

- A new package: `@tinacms/mdx`
- Updates to the rich-text editor
  - Mainly updates to handling code blocks, raw html, and invalid markdown

The best way to review this PR is to just play with the rich text editor and possibly run some of the tests in the `@tinacms/mdx` package. Those tests are pretty detailed and I've created a video to walk through how to work with them. https://www.loom.com/share/d3a8dd3235594647982fe3a939a661bf

closes #2884
closes #2881

### CODE REVIEW

File: packages/@tinacms/mdx/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,1104 @@
+# tina-graphql
```
Comment (kldavis4) [inline packages/@tinacms/mdx/CHANGELOG.md:1]: this looks like the changelog from @tinacms/mdx - is that intended?

File: packages/@tinacms/mdx/package.json [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  },
+  "repository": {
+    "url": "https://github.com/tinacms/tinacms.git",
+    "directory": "packages/tina-graphql"
+  },
+  "devDependencies": {
+    "@swc/core": "^1.2.210",
+    "@swc/jest": "^0.2.21",
+    "@tinacms/datalayer": "workspace:*",
+    "@tinacms/scripts": "workspace:*",
```
Comment (kldavis4) [inline packages/@tinacms/mdx/package.json:52]: are these dependencies needed for this package?

File: packages/@tinacms/mdx/src/tests/autotest/shortcodes.test.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    },
+    {
+      type: 'p',
+      children: [
+        {
+          type: 'mdxJsxTextElement',
+          name: 'Shortcode1',
+          children: [{ type: 'text', text: '' }],
+          props: {
+            text: 'rimg src="https://res.cloudinary.com/indysigner/image/upload/v1609336455/the-relevance-gap_kzwi6q.png" href="https://infrequently.org/2020/06/platform-adjacency-theory/" sizes="100vw" caption="Image credit: <a href=\'https://infrequently.org/2020/06/platform-adjacency-theory/\'>Alex Russell</a>" alt="The Relevance Gap"',
```
[the remark this answers is not in the document]
Comment (jeffsee55) [reply packages/@tinacms/mdx/src/tests/autotest/shortcodes.test.ts:44]: This is an example from a hugo site, no idea what `rimg` is but the idea is that it's treated as plaintext from the editor

**Comment by kldavis4** [comment]
@jeffsee55 I am getting this error when trying to load blog posts in tina-cloud-starter in the monorepo:

```
Error querying file content/posts/anotherPost.mdx collection posts. Please run "tinacms audit" or add the --verbose option for more info

GraphQL request:8:11
7 |           templates
8 |           documents(sort: $sort, after: $after, first: $limit) @include(if: $includeDocuments) {
  |           ^
9 |             totalCount
More error context below
Error querying file content/posts/anotherPost.mdx collection posts. Please run "tinacms audit" or add the --verbose option for more info
TypeError: preprocessedString.replaceAll is not a function
    at /Users/kldavis/projects/tinacms/packages/@tinacms/mdx/dist/index.js:244677:47
    at Array.forEach (<anonymous>)
    at markdownToAst (/Users/kldavis/projects/tinacms/packages/@tinacms/mdx/dist/index.js:244672:66)
    at parseMDX (/Users/kldavis/projects/tinacms/packages/@tinacms/mdx/dist/index.js:244695:12)
    at Resolver.resolveFieldData (/Users/kldavis/projects/tinacms/packages/@tinacms/graphql/dist/index.js:3400:48)
    at /Users/kldavis/projects/tinacms/packages/@tinacms/graphql/dist/index.js:3038:25
    at reducePromises (/Users/kldavis/projects/tinacms/packages/@tinacms/graphql/dist/index.js:95:12)
    at async sequential (/Users/kldavis/projects/tinacms/packages/@tinacms/graphql/dist/index.js:97:18)
    at async Resolver.getDocument (/Users/kldavis/projects/tinacms/packages/@tinacms/graphql/dist/index.js:3037:11)
    at async /Users/kldavis/projects/tinacms/packages/@tinacms/graphql/dist/index.js:4040:26 {
  message: 'Error querying file content/posts/anotherPost.mdx collection posts. Please run "tinacms audit" or add the --verbose option for more info',
  locations: [ { line: 8, column: 11 } ],
  path: [ 'collection', 'documents' ]
}
```

Looks like its calling replaceAll in the backend

**Comment by jeffsee55** [comment]
> Looks like its calling replaceAll in the backend

Oh that's only in node 15 > 🤔 . Wonder why this past in CI. Thanks.

File: packages/@tinacms/mdx/src/parse/index.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+ * While it may be desirable to ultimately serve a remark AST shape as part of the API response,
+ * it's currently much easier to only support the shape that works with Slate. This is ok for now for 2
+ * reasons.
+ *
+ * 1. Us providing the `TinaMarkdown` component on the frontend abstracts away an work the developer
+ * would need to do, so it doesn't really matter what shape the response is as long as the external API
+ * doesn't change
+ *
+ * 2. We don't need to do any client-side parsing. Since TinaMarkdown and the slate editor work with the same
+ * format we can just allow Tina to do it's thing and update the form valuse with no additional work.
```
Comment (kldavis4) [inline packages/@tinacms/mdx/src/parse/index.ts:72]:
```suggestion
 * format we can just allow Tina to do it's thing and update the form value with no additional work.
```

**Review by kldavis4 — APPROVED** [review]
went through all the test cases and they make sense (after clarifying questions)

**Review by kldavis4 — APPROVED** [review]

**Pull request merged**